OpenSim/Actuators/Millard2012EquilibriumMuscle.cpp

```diff
@@ -1364,122 +1281,181 @@ Millard2012EquilibriumMuscle::estimateMuscleFiberState(
-        dFmAT_dlceAT = calc_DFiberForceAT_DFiberLengthAT(dFmAT_dlce,sinphi,
-                                                         cosphi,lce);
-        dFt_d_tl     = fseCurve.calcDerivative(tlN,1)*fiso/tsl;
-        dFt_d_lce    = calc_DTendonForce_DFiberLength(dFt_d_tl,lce,
-                                                      sinphi,cosphi);
-
-        // Error derivative
-        dferr_d_lce = dFmAT_dlce - dFt_d_lce;
+            // Update fiber velocity
+            dlce = getPennationModel().calcFiberVelocity(cosphi, dml, dtl);
+            dlceN = dlce / (vmax*ofl);
+            // Update the force-velocity multiplier
+            fv = get_ForceVelocityCurve().calcValue(dlceN);
+        }
+    };
 
-        if(abs(ferr) > aSolTolerance) {
-            if(abs(dferr_d_lce) > SimTK::SignificantReal) {
-                // Take a full Newton Step if the derivative is nonzero
-                delta_lce = -ferr/dferr_d_lce;
-                lce       = lce + delta_lce;
-            } else {
-                // We've stagnated; perturb the current solution
-                double perturbation =
-                    2.0*((double)rand())/((double)RAND_MAX)-1.0;
-                double lengthPerturbation =
-                    0.5*perturbation*getOptimalFiberLength();
-                lce += lengthPerturbation;
+    //*******************************
+    //Initialize the loop
+    int iter = 0;
+
+    // Estimate the position level quantities (lengths, angles) of the muscle
+    positionFunc();
+
+    // Multipliers based on initial fiber-length estimate
+    multipliersFunc();
+
+    // Starting guess at the force-velocity multiplier is static
+    fv = 1.0;
+
+    fiberForceV = calcFiberForce(fiso, ma, fal, fv, fpe, dlceN);
+    Fm = fiberForceV[0];
+
+    // Compute the partial derivative of the force error w.r.t. lce
+    partialsFunc();
+
+    // update fiber and tendon velocity and velocity multiplier
+    velocityFunc();
+
+    // Compute the force error
+    ferrFunc();
+
+    // Update the partial derivatives of the force error w.r.t. lce with 
+    // newly estimated fv
+    partialsFunc();
+
+    double ferrPrev = ferr;
+    double lcePrev = lce;
+
+    double h =1.0;
+    while( (abs(ferr) > aSolTolerance) && (iter < aMaxIterations)) {
+        // Compute search direction
+        dferr_d_lce = dFmAT_dlce - dFt_d_lce;
+        h = 1.0;
+
+        while (abs(ferr) >= abs(ferrPrev)) {
+            // Compute the Newton step
+            delta_lce = -h*ferrPrev / dferr_d_lce;
+            // Take a Newton Step if the step is nonzero
+            if (abs(delta_lce) > SimTK::SignificantReal)
+                lce = lcePrev + delta_lce;
+            else {
+                // We've stagnated or hit a limit; assume we are hitting local
+                // minimum and attempt to approach from the other direction.
+                lce = lcePrev - sign(delta_lce)*SimTK::SqrtEps;
```

Comment:
It seems to work for the Thelen2003Muscle. I haven't been able to get it to fail since I added this. From plotting the ferr curve as a function of lce, I was able to see that the solution was close, but not able to converge because of big changes in curvature (from very flat to steep or vice versa), in which case approaching from the other side seems to help. Furthermore, the curve changes with different fv, so re-evaluating the derivatives and fv helps. If we are not updating fv, it doesn't hit this condition.